### PR TITLE
Feature/transitive brands

### PR DESCRIPTION
**Pull request by khatton-ft**
- Add support for `PUT` requests.
- Remove existing `implicitlyClassifiedBy` annotations, and add `implicitlyClassifiedBy` by traversing the brands hierarchy on a `PUT` request, returning the result in the response body. Annotations for genres are excluded from the above logic, but other concept types for `isClassifiedBy` annotations (which are not expected from TagMe) will result in excess 404 requests to the brands API.

### CODE REVIEW

File: annotations/handler.go [hunk 1]
```diff
@@ -6,59 +6,67 @@ import (
...
-		http.Error(w, err.Error(), http.StatusInternalServerError)
+		switch err.(type) {
+		case *UPPAnnotationsApiError:
+			apiError := err.(*UPPAnnotationsApiError)
+			if apiError.StatusCode == http.StatusNotFound || apiError.StatusCode == http.StatusBadRequest {
+				w.WriteHeader(apiError.StatusCode)
+				io.Copy(w, apiError.Body)
+			} else {
+				log.WithError(err).WithField(tidutils.TransactionIDKey, tID).WithField("uuid", uuid).Error("Error in reading annotations")
+				writeMessage(w, "Service unavailable", http.StatusServiceUnavailable)
```
[the remark this answers is not in the document]
Comment (khatton-ft) [reply annotations/handler.go]: No, if UPP returns anything that is not a 200, 400 or 404, we consider that to mean it's unavailable.

File: annotations/implications_test.go [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+
+	parentBrandUuid := uuid.NewV4().String()
+	parentBrandId := "http://www.ft.com/thing/" + parentBrandUuid
+
+	brandsResolver := &mockBrandsResolver{}
+	brandsResolver.On("GetBrands", ann1.ConceptId).Return([]string{ann1.ConceptId, parentBrandId}, nil)
+
+	ann2 := mentionsAnnotation()
+	ann3 := assertedBrandAnnotation()
+
```
[the remark this answers is not in the document]
Comment (khatton-ft) [reply annotations/implications_test.go]: By calling `AssertExpectations` at line 207, the test case proves that there was no call to the brands resolver for `ann3`, so it was excluded.

File: annotations/brandsResolver_test.go [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+
+func (s *BrandsResolverTestSuite) TestGetFtChildBrandUsingParent() {
+	i := strings.LastIndex(fastFtBrand, "/")
+	s.brands[fastFtBrand[i+1:]] = Brand{ID: fastFtBrand, ParentBrand: &Brand{ID: ftBrand}}
+	server := httptest.NewServer(s.handler)
+	defer server.Close()
+
+	resolver := NewBrandsResolver(server.URL + "/%v", testAPIKey, s.linter)
+
+	actual := resolver.GetBrands(fastFtBrand)
```
[the remark this answers is not in the document]
Comment (khatton-ft) [reply annotations/brandsResolver_test.go]: No, but it's using the fact that fastFtBrand has a parent brand of ftBrand (line 84), as opposed to the next test case, which uses the fact that ftBrand has a child brand of fastFtBrand (line 98).

[the remark this answers is not in the document]
Comment (galiri) [reply annotations/brandsResolver_test.go]: yes it does, to my mind it's clearer to leave the  data the same and change the call, but that is doing the right thing

**Comment by khatton-ft** [comment]
We have decided not to do this on save of draft annotations.

**Pull request closed without merging**